letsencrypt/acme/messages.py

```diff
@@ -74,35 +37,58 @@ def get_msg_cls(cls, jobj):
         try:
             msg_cls = cls.TYPES[msg_type]
         except KeyError:
-            raise errors.UnrecognizedMessageTypeError(msg_type)
+            raise errors.UnrecognizedTypeError(msg_type)
 
         return msg_cls
 
     @classmethod
-    def from_json(cls, jobj, validate=True):
-        """Deserialize validated ACME message from JSON string.
+    def from_json(cls, jobj):
+        """Deserialize from (possibly invalid) JSON object.
+
+        Note that the input ``jobj`` has not been sanitized in any way.
 
-        :param str jobj: JSON object.
-        :param bool validate: Validate against schema before deserializing.
-            Useful if :class:`JWK` is part of already validated json object.
+        :param jobj: JSON object.
 
-        :raises letsencrypt.acme.errors.ValidationError: if validation
-            was unsuccessful
+        :raises letsencrypt.acme.errors.SchemaValidationError: if ``validate``
+            was ``True`` and object couldn't be validated.
 
-        :returns: Valid ACME message.
-        :rtype: subclass of :class:`Message`
+        :returns: instance of the class
 
         """
         msg_cls = cls.get_msg_cls(jobj)
-        if validate:
-            msg_cls.validate_json(jobj)
-        # pylint: disable=protected-access
-        return msg_cls._from_valid_json(jobj)
+
+        try:
+            jsonschema.validate(jobj, msg_cls.schema)
+        except jsonschema.ValidationError as error:
+            raise errors.SchemaValidationError(error)
+
+        return cls.from_valid_json(jobj)
+
+    @classmethod
+    def json_loads(cls, json_string):
+        """Load JSON string."""
+        return cls.from_json(json.loads(json_string))
+
+    def json_dumps(self, *args, **kwargs):
+        """Dump to JSON string using proper serializer.
+
+        :returns: JSON serialized string.
+        :rtype: str
+
+        """
+        return json.dumps(
+            self, *args, default=util.dump_ijsonserializable, **kwargs)
```

Comment:
util.dump_ijsonserializable - I guess I don't understand why this is necessary, if only json_dumps in TypedACMEObject uses it... and TypedACMEObject is of type ACMEObject which implements IJSONSerialiable.  Is there some other python_objects being passed to it that I missed or why is this necessary.

I see some explanation in acme/interfaces.py about returning different objects.  I didn't immediately see anything... is there a specific reference?